x/ccv/provider/keeper/key_assignment.go

```diff
@@ -0,0 +1,739 @@
+	return ka.Store.GetConsumerPublicKeyToProviderPublicKey(ck)
+}
+
+// GetProviderPubKeyFromConsumerConsAddress returns the provider key which was associated to the consumer key with
+// consensus address cca at the last update that used the consumer key.
+func (ka *KeyAssignment) GetProviderPubKeyFromConsumerConsAddress(cca sdk.ConsAddress) (pk ProviderPublicKey, found bool) {
+	if lum, found := ka.Store.GetConsumerConsAddrToLastUpdateMemo(cca); found {
+		return *lum.ProviderKey, true
+	}
+	return pk, false
+}
+
+// PruneUnusedKeys deletes from the store all of the consumer keys which the consumer chain has matured, based
+// on latestMatureVscid, and for which it sure that the correct consumer chain can no longer reference the key
+// in a downtime or double sign slash instruction.
+func (ka *KeyAssignment) PruneUnusedKeys(latestMatureVscid VSCID) {
+	toDel := []ConsumerConsAddr{}
+	ka.Store.IterateConsumerConsAddrToLastUpdateMemo(func(cca ConsumerConsAddr, lum providertypes.LastUpdateMemo) bool {
+		if lum.Power == 0 && lum.Vscid <= latestMatureVscid {
+			toDel = append(toDel, cca)
+		}
+		return false
+	})
+	for _, cca := range toDel {
+		ka.Store.DelConsumerConsAddrToLastUpdateMemo(cca)
+	}
+}
+
+// DeterministicStringify returns a deterministic string representation of the
+// key. This is useful to identify like keys with different representations.
+func DeterministicStringify(k tmprotocrypto.PublicKey) string {
+	bz, err := k.Marshal()
+	if err != nil {
+		panic(err)
+	}
+	return string(bz)
+}
+
+// KeyToPower is an implementation of a map from a public key to a power.
+// It is used because public keys are not comparable in Go, so they cannot be
+// used as builtin map keys.
+type KeyToPower struct {
+	inner     map[tmprotocrypto.PublicKey]int64
+	canonical map[string]tmprotocrypto.PublicKey
+}
+
+func MakeKeyToPower() KeyToPower {
+	return KeyToPower{
+		inner:     map[tmprotocrypto.PublicKey]int64{},
+		canonical: map[string]tmprotocrypto.PublicKey{},
+	}
+}
+
+// Set is a typical map set operation.
+func (m *KeyToPower) Set(pk tmprotocrypto.PublicKey, power int64) {
+	stringified := DeterministicStringify(pk)
+	if k, found := m.canonical[stringified]; found {
+		pk = k
+	} else {
+		m.canonical[stringified] = pk
+	}
+	m.inner[pk] = power
+}
+
+// Get is a typical map get operation.
+func (m *KeyToPower) Get(pk tmprotocrypto.PublicKey) (int64, bool) {
+	if k, found := m.canonical[DeterministicStringify(pk)]; found {
+		power, found := m.inner[k]
+		if !found {
+			panic("found canonical key but key not present in inner map")
+		}
+		return power, true
+	}
+	return -1, false
+}
+
+// KeyToLastUpdateMemo is an implementation of a map from a public key to a
+// memo storing the last update.
+// It is used because public keys are not comparable in Go, so they cannot be
+// used as builtin map keys.
+type KeyToLastUpdateMemo struct {
```

Comment:
These types all share an inner and canonical map, along with some shared methods. Could we use go generics to reduce the amount of code that is introduced?

Reply:
Yeah I'm happy to use generics. I didn't want to bust them out yet because

1. I've not used golang generics before (although I've used them plenty in other languages)
2. We only just upgraded to 1.18, on a technicality, the code is really written in 1.17 style. We have currently have 0 generics and I think most cosmos code is written without generics AFAIU
3. I'm lazy and this didn't seem like a big deal, it's only 100 lines. Trying to KISS

I wouldn't mind using generics tbh. If you feel strongly I can do it